src/addon/addoneventpropagator.js

```diff
@@ -0,0 +1,156 @@
+/*globals requireJS*/
+/*jshint node:true*/
+/**
+ * This propagates the events from the storage to the addon-handler process.
+ *
+ * The addon-handler can take two forms:
+ * - If url is provided it will post requests to a separate server see (bin/addon_handler.js)
+ * - Otherwise it will send messages to a connected worker.
+ *
+ * @author pmeijer / https://github.com/pmeijer
+ */
+'use strict';
+
+var superagent = require('superagent'),
+    Q = require('q'),
+    CONSTANTS = requireJS('common/Constants'),
+    WORKER_CONSTANTS = require('../server/worker/constants');
+
+function AddOnEventPropagator(storage, serverWorkerManager, mainLogger, gmeConfig) {
+
+    var connectedWorker = !gmeConfig.addOn.workerUrl,
+        logger = mainLogger.fork('AddOnEventPropagator'),
+        statusUrl,
+        COPY = function (obj) {
+            return JSON.parse(JSON.stringify(obj));
+        };
+
+    function branchJoined(_s, data) {
```

Comment:
this and the updated function are almost identical, it might would worth refactoring them into a single function.